newspeek/src/main/java/use_case/helpers/CharByCharCensorshipService.java

```diff
@@ -15,18 +17,29 @@ public Article censor(Article article, CensorshipRuleSet ruleset) {
         Article result = article.copy();
         StringBuilder censoredText = new StringBuilder();
         StringBuilder word = new StringBuilder();
-
+        int censoredWordsCount = 0;
+        int replacedWordsCount = 0;
         for (char character : article.getText().toCharArray()) {
             if (PUNCTUATION.indexOf(character) != -1) {
                 // character is punctuation
                 censoredText.append(censorWord(word.toString(), ruleset));
                 censoredText.append(character);
+                if (!Objects.equals(censorWord(word.toString(), ruleset), word.toString())) {
```

Comment:
This is checking if the strings are the same object.

Also, this makes many calls to censorWord, which requires an unnecessary number of map lookups